oap-server/server-storage-plugin/storage-iotdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/iotdb/query/IoTDBTopologyQueryDAO.java

```diff
@@ -0,0 +1,255 @@
+        try {
+            wrapper = sessionPool.executeQueryStatement(query.toString());
+            if (log.isDebugEnabled()) {
+                log.debug("SQL: {}, columnNames: {}", query, wrapper.getColumnNames());
+            }
+            while (wrapper.hasNext()) {
+                RowRecord rowRecord = wrapper.next();
+                List<Field> fields = rowRecord.getFields();
+                Call.CallDetail call = new Call.CallDetail();
+                String[] layerNames = fields.get(0).getStringValue().split("\\" + IoTDBClient.DOT + "\"");
+                String entityId = client.layerName2IndexValue(layerNames[2]);
+                final int componentId = fields.get(1).getIntV();
+                call.buildFromInstanceRelation(entityId, componentId, detectPoint);
+                calls.add(call);
+            }
+        } catch (IoTDBConnectionException | StatementExecutionException e) {
+            throw new IOException(e);
+        } finally {
+            sessionPool.closeResultSet(wrapper);
+        }
+        return calls;
+    }
+
+    private List<Call.CallDetail> loadEndpointFromSide(String tableName, long startTB, long endTB,
+                                                       String sourceCName, String destCName,
+                                                       String id, boolean isSourceId) throws IOException {
+        // This method don't use "group by" like other storage plugin.
+        StringBuilder query = new StringBuilder();
+        query.append("select * from ");
+        query = client.addModelPath(query, tableName);
+        query = client.addQueryAsterisk(tableName, query);
+        query.append(" where ").append(IoTDBClient.TIME).append(" >= ").append(TimeBucket.getTimestamp(startTB))
+                .append(" and ").append(IoTDBClient.TIME).append(" <= ").append(TimeBucket.getTimestamp(endTB));
+        query.append(" and ").append(isSourceId ? sourceCName : destCName).append(" = \"").append(id).append("\"")
+                .append(IoTDBClient.ALIGN_BY_DEVICE);
+
+        SessionPool sessionPool = client.getSessionPool();
+        SessionDataSetWrapper wrapper = null;
+        List<Call.CallDetail> calls = new ArrayList<>();
+        try {
+            wrapper = sessionPool.executeQueryStatement(query.toString());
+            if (log.isDebugEnabled()) {
+                log.debug("SQL: {}, columnNames: {}", query, wrapper.getColumnNames());
+            }
+            while (wrapper.hasNext()) {
+                RowRecord rowRecord = wrapper.next();
+                List<Field> fields = rowRecord.getFields();
+                Call.CallDetail call = new Call.CallDetail();
+                String[] layerNames = fields.get(0).getStringValue().split("\\" + IoTDBClient.DOT + "\"");
+                String entityId = client.layerName2IndexValue(layerNames[2]);
+                // TODO check DetectPoint
+                call.buildFromEndpointRelation(entityId, DetectPoint.SERVER);
```

Comment:
![image](https://user-images.githubusercontent.com/5441976/143200217-deced04a-d6b5-491f-a3bf-279742a4aebd.png)

This is client side detecting, what do you mean exactly?

Reply:
> This is client side detecting, what do you mean exactly?

No, it's not this method. In line 291 in `TopologyQueryEsDAO` and line 270 in `H2TopologyQueryDAO`, the second parameter of `buildFromEndpointRelation` always is `DetectPoint.SERVER`. But in line 305 of `TopologyQuery`, it can be `DetectPoint.SERVER` or `DetectPoint.CLIENT`

Reply:
For endpoint, yes. The dependency is only `server` side detection.